crates/teloxide/examples/composite_state.rs

```diff
@@ -0,0 +1,190 @@
+//! }
+//!
+//! // More enumeration definitions...
+//! ```
+//!
+//! This example demonstrates how to achieve this `teloxide` design pattern.
+
+use teloxide::{
+    dispatching::{dialogue::InMemStorage, MessageFilterExt},
+    prelude::*,
+    types::{ChatId, Message},
+};
+
+type Bot = teloxide::Bot;
+type Error = Box<dyn std::error::Error + Send + Sync>;
+type HandlerResult = Result<(), Error>;
+type UpdateHandler = teloxide::dispatching::UpdateHandler<Error>;
+type Storage = InMemStorage<GlobalState>;
+type Dialogue = teloxide::dispatching::dialogue::Dialogue<GlobalState, Storage>;
+
+#[derive(BotCommands, Clone)]
+#[command(rename_rule = "snake_case")]
+enum Command {
+    #[command(description = "Setup user account")]
+    Start,
+}
+
+#[derive(Clone, Default)]
+enum GlobalState {
+    #[default]
+    Unconfigured,
+    UserSetup(UserSetup),
+    Idle,
+}
+
+#[derive(Clone)]
+enum UserSetup {
+    ReceiveFullName,
+    ReceiveAge { full_name: String },
+}
+
+/// Helper struct to store only required information to answer messages and
+/// reduce the size of the stack frames for handler functions
+#[derive(Clone)]
+struct IdsBundle {
+    chat_id: ChatId,
+    // Can be used to reply to messages or edit messages
+    // message_id: MessageId,
+}
+
+#[tokio::main]
+async fn main() {
+    pretty_env_logger::init();
+    log::info!("Starting the \"composite_state\" example");
+
+    let bot = Bot::from_env();
+
+    Dispatcher::builder(bot, schema())
+        .dependencies(dptree::deps![InMemStorage::<GlobalState>::new()])
+        .build()
+        .dispatch()
+        .await
+}
+
+fn schema() -> UpdateHandler {
+    Update::filter_message()
+        /*
+           Currently the size of the `Message` struct (for TBA 6.9) is 1936 bytes, it's insane to copy it entirely in every handler's stack.
+           So, here I introduce the `IdsBundle` which is 8 bytes in size, because all we need is a `chat_id`.
+           The similar thing can be applied to the `CallbackQuery` struct which is
+           even bigger..
+           Take a look at this issue: https://github.com/teloxide/teloxide/issues/1118, maybe there will be
+           more appropriate approach: `Arc<Message>` or similar.
+        */
+        .map(|msg: Message| IdsBundle { chat_id: msg.chat.id })
+        .branch(
+            Message::filter_text()
+                .enter_dialogue::<Message, Storage, GlobalState>()
+                .branch(
+                    teloxide::filter_command::<Command, _>()
+                        .branch(dptree::case![Command::Start].endpoint(ask_full_name)),
+                )
+                .branch(dptree::case![GlobalState::Idle].endpoint(handle_configured_user_message))
+                .branch(
+                    /*
+                       Its essential not to use `dptree::case![GlobalState::UserSetup(UserSetup::ReceiveFullName)]` directly, this won't work.
+
+                       Each nested enum requires it's own `branch` scope.
```

Comment:
I guess this syntax will simply not compile? Why discuss it at all?

Reply:
It arises from time to time in discussions, so some users think it's not possible at all